### PR TITLE
add azure openai api support to ontogpt

### PR DESCRIPTION
**Pull request by k-bandi**
This PR adds support for using Azure OpenAI API endpoints in OntoGPT for the LLM component instead of Vanilla OpenAI endpoints.

### CODE REVIEW

File: src/ontogpt/clients/openai_client.py [hunk 1]
```diff
@@ -28,10 +31,18 @@ def __post_init__(self):
         if not self.api_key:
             self.api_key = get_apikey_value("openai")
         openai.api_key = self.api_key
+        self.client = AzureOpenAI(
+            # https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#rest-api-versioning
```
Comment (lucinvitae) [inline src/ontogpt/clients/openai_client.py]: remove

File: src/ontogpt/clients/openai_client.py [hunk 2]
```diff
@@ -28,10 +31,18 @@ def __post_init__(self):
...
+            api_version=AZURE_API_VERSION,
+            # https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource?pivots=web-portal#create-a-resource
```
Comment (lucinvitae) [inline src/ontogpt/clients/openai_client.py]: remove

**Review by lucinvitae — APPROVED** [review]

File: pyproject.toml [hunk 3]
```diff
@@ -2,60 +2,62 @@
 name = "ontogpt"
 version = "0.3.7"
 description = "OntoGPT"
-authors = ["Chris Mungall <cjmungall@lbl.gov>"]
+authors = ["Chris Mungall <cjmungall@lbl.gov>", "Krishna Bandi <krishna.bandi@invitae.com>"]
 license = "BSD-3"
 readme = "README.md"
 
 [tool.poetry.dependencies]
```
Comment (lucinvitae) [inline pyproject.toml]: consider adding a todo about hardcoded deps while we sort out `poetry lock` runtime issues.

File: src/ontogpt/clients/openai_client.py [hunk 4]
```diff
@@ -28,10 +31,18 @@ def __post_init__(self):
...
+            azure_endpoint=AZURE_ENDPOINT,
+            api_key=self.api_key,
+            azure_deployment=AZURE_MODEL,
+        )
 
     # TODO: Dynamically update max_tokens
-    def complete(self, prompt, max_tokens=3000, show_prompt: bool = False, **kwargs) -> str:
-        engine = self.model
+    def complete(self, prompt, max_tokens=500, show_prompt: bool = False, **kwargs) -> str:
+        engine = AZURE_MODEL
```
Comment (lucinvitae) [inline src/ontogpt/clients/openai_client.py:45]:
```suggestion
        # TODO: dynamically set model at call time
        engine = AZURE_MODEL
```

File: src/ontogpt/clients/openai_client.py [hunk 5]
```diff
@@ -28,10 +31,18 @@ def __post_init__(self):
         if not self.api_key:
             self.api_key = get_apikey_value("openai")
         openai.api_key = self.api_key
+        self.client = AzureOpenAI(
```
Comment (lucinvitae) [inline src/ontogpt/clients/openai_client.py:35]:
```suggestion
        # TODO: control client (Azure vs not) using a feature flag
        self.client = AzureOpenAI(
```

File: src/ontogpt/clients/openai_client.py [hunk 6]
```diff
@@ -63,11 +74,7 @@ def complete(self, prompt, max_tokens=3000, show_prompt: bool = False, **kwargs)
                         **kwargs,
                     )
                 else:
-                    response = openai.Completion.create(
-                        engine=engine,
-                        prompt=prompt,
-                        max_tokens=max_tokens,
-                    )
+                    raise ValueError("Unsupported mode")
```
Comment (lucinvitae) [inline src/ontogpt/clients/openai_client.py:78]: As per deprecation notices ([1](https://platform.openai.com/docs/guides/text-generation/completions-api), [2](https://openai.com/blog/gpt-4-api-general-availability)):
```suggestion
                    # TODO: remove chat api flag and eliminate references to legacy completions API
                    raise ValueError("Unsupported mode")
```

**Review by lucinvitae — COMMENTED** [review]
We should add a few todos here

File: src/settings.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+import dpath
+import toml
+
+BASE_DIR = Path(__file__).parent.parent.absolute()
+
+
+def is_test() -> bool:
+    """Returns true if running in a test environment."""
+    return "pytest" in sys.argv[0]
```
Comment (lucinvitae) [inline src/settings.py]: this can be removed since it's specific to our test suites.

File: src/settings.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def read_toml(path: Path) -> Dict[str, Any]:
+    with path.open() as cf:
+        return toml.load(cf)
+
+
+def merge_settings(overrides: Dict[str, Any], defaults: Dict[str, Any]) -> None:
```
Comment (lucinvitae) [inline src/settings.py:16]: remove

File: src/settings.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    if settings_path.is_file():
+        settings = read_toml(settings_path)
+    else:
+        settings = {}
+
+    settings_override_file = etc_dir / f"{settings_name}_custom.toml"
+    if settings_override_file.is_file() and not is_test():
+        overrides = read_toml(settings_override_file)
+        merge_settings(overrides, settings)
```
Comment (lucinvitae) [inline src/settings.py]: simplify
```suggestion
    settings_path = os.getenv("SETTINGS_PATH", None)
    settings_name = os.getenv("SETTINGS_NAME", "local")

    etc_dir = base_dir / "etc"
    default_settings = etc_dir / f"{settings_name}.toml"
    settings_override = etc_dir / f"{settings_name}_custom.toml"

    if settings_path:
        settings_path = Path(settings_path)
    elif settings_override.is_file():
        settings_path = settings_override
    else:
        settings_path = default_settings

    settings = read_toml(settings_path) if settings_path.is_file() else {}
```

File: pyproject.toml (hunk 3 above)
Comment (k-bandi) [reply pyproject.toml]: adding urllib3 solved the poetry lock time issue (it is around a minute now).

**Pull request merged**